### PR TITLE
Add scope validator for application management core

### PR DESCRIPTION
**Pull request by senthalan**

### CODE REVIEW

**Comment by senthalan** [comment]
depend on script changes of test cases #809

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java [hunk 1]
```diff
@@ -675,7 +688,7 @@ public OAuthRevocationResponseDTO revokeAuthzForAppsByResoureOwner(
                     if (appDTO.getApplicationName().equals(appName)) {
                         Set<AccessTokenDO> accessTokenDOs;
                         try {
-                            // retrieve all ACTIVE or EXPIRED access tokens for particular client authorized by this user
+                            //retrieve all ACTIVE or EXPIRED access tokens for particular client authorized by this user
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java]: You should have a space before starting the comment. Is there any specific reason to remove this space?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 2]
```diff
@@ -126,25 +139,41 @@ public void addOAuthApplication(OAuthAppDO consumerAppDO) throws IdentityOAuthAd
...
                         prepStmt.execute();
+                        try (ResultSet results = prepStmt.getGeneratedKeys()) {
+                            if (results.next()) {
+                                appId = results.getInt(1);
+                            }
+                        }
                     }
                 }
-
+                // some JDBC Drivers returns this in the result, some don't
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Comments should start with capital letter.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 3]
```diff
@@ -126,25 +139,41 @@ public void addOAuthApplication(OAuthAppDO consumerAppDO) throws IdentityOAuthAd
...
+                if (appId == 0) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("JDBC Driver did not return the app id, executing Select operation");
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Change the debug log to more understandable way

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 4]
```diff
@@ -770,6 +827,115 @@ private void removeOauthOIDCPropertyTable(Connection connection,
...
+            throws SQLException, InvalidOAuthClientException, IdentityOAuth2Exception {
+
+        int appId = 0;
+        try (PreparedStatement prepStmt = connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries
+                .GET_APP_ID_BY_CONSUMER_KEY)) {
+            prepStmt.setString(1, persistenceProcessor.getProcessedClientId(clientId));
+            try (ResultSet rSet = prepStmt.executeQuery()) {
+                boolean rSetHasRows = false;
+                while (rSet.next()) {
+                    // There is at least one application associated with a given key
```
Comment (nilasini) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: Comments should end with a full stop. Please check in other places as well.

Comment (senthalan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/942/commits/ca4690d734709b108ef2df33a69ae3c9242febc8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 3 above)
Comment (senthalan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/942/commits/ca4690d734709b108ef2df33a69ae3c9242febc8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 2 above)
Comment (senthalan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/942/commits/ca4690d734709b108ef2df33a69ae3c9242febc8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java (hunk 1 above)
Comment (senthalan) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/942/commits/ca4690d734709b108ef2df33a69ae3c9242febc8

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java [hunk 5]
```diff
@@ -876,9 +909,30 @@ private void validateTokenExpiryConfigurations(OAuthConsumerAppDTO oAuthConsumer
...
-            log.debug("Invalid expiry time value '0' set for " + tokenType + " in ServiceProvider: " + appName + ". " +
-                    "Defaulting to expiry value: " + defaultValue + " seconds.");
+            log.debug("Invalid expiry time value '0' set for " + tokenType + " in ServiceProvider: " + appName + ". "
+                    + "Defaulting to expiry value: " + defaultValue + " seconds.");
         }
     }
-}
 
+    /**
+     *  Get the scope validators registered by the user and filter the allowed ones.
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java]: Who is this user? comment is not clear

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java [hunk 6]
```diff
@@ -876,9 +909,30 @@ private void validateTokenExpiryConfigurations(OAuthConsumerAppDTO oAuthConsumer
...
+     * @throws IdentityOAuthAdminException  Identity OAuthAdmin exception.
+     */
+    private String[] filterScopeValidators(OAuthConsumerAppDTO application) throws IdentityOAuthAdminException {
+
+        List<String> scopeValidators = new ArrayList<>(Arrays.asList(getAllowedScopeValidators()));
+        String[] requestedScopeValidators = application.getScopeValidators();
+        if (requestedScopeValidators == null) {
+            requestedScopeValidators = new String[0];
+        }
+        for (String requestedScopeValidator : requestedScopeValidators) {
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java]: `        for (String requestedScopeValidator : requestedScopeValidators) {
            if (!scopeValidators.contains(requestedScopeValidator)) {
                throw new IdentityOAuthAdminException(requestedScopeValidator + " not allowed");
            }
        }`
Will execute only if requestedScopeValidators != null

File: components/org.wso2.carbon.identity.oauth/src/test/resources/dbScripts/h2_with_application_and_token.sql [hunk 7]
```diff
@@ -20,6 +20,13 @@ CREATE TABLE IF NOT EXISTS IDN_OAUTH_CONSUMER_APPS (
             PRIMARY KEY (ID)
 );
 
+CREATE TABLE IF NOT EXISTS IDN_OAUTH2_SCOPE_VALIDATORS (
+	APP_ID INTEGER NOT NULL,
+	SCOPE_VALIDATOR VARCHAR (128) NOT NULL,
+	PRIMARY KEY (APP_ID,SCOPE_VALIDATOR),
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/test/resources/dbScripts/h2_with_application_and_token.sql]: Formatting error

File: components/org.wso2.carbon.identity.oauth/src/test/resources/dbScripts/identity.sql [hunk 8]
```diff
@@ -27,6 +27,13 @@ CREATE TABLE IF NOT EXISTS IDN_OAUTH_CONSUMER_APPS (
             PRIMARY KEY (ID)
 );
 
+CREATE TABLE IF NOT EXISTS IDN_OAUTH2_SCOPE_VALIDATORS (
+	APP_ID INTEGER NOT NULL,
+	SCOPE_VALIDATOR VARCHAR (128) NOT NULL,
+	PRIMARY KEY (APP_ID,SCOPE_VALIDATOR),
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/test/resources/dbScripts/identity.sql]: Formatting issue

**Pull request merged**